### PR TITLE
Replace sparkle with glow

### PR DESCRIPTION
**Pull request by sagudev**
[`sparkle`](https://github.com/servo/sparkle) is a GL crate that is part of the Servo project, but mostly unmaintained. We would like to archive it and replace it with [`glow`](https://github.com/grovesNL/glow/tree/main), that is actively maintained and used by `wgpu` and the wider Rust community.

Part of https://github.com/servo/servo/issues/33539

### CODE REVIEW

**Comment by mrobinson** [comment]
Nice! I think it would be really useful here to give a bit of background on `sparkle` and `glow` in the PR description and why we are replacing one for the other.

File: src/chains.rs [hunk 1]
```diff
@@ -191,23 +197,31 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
...
-                0,
-                back_info.size.width,
-                back_info.size.height,
-                gl::COLOR_BUFFER_BIT | gl::DEPTH_BUFFER_BIT | gl::STENCIL_BUFFER_BIT,
-                gl::NEAREST,
-            );
-            debug_assert_eq!(gl.get_error(), gl::NO_ERROR);
+            unsafe {
+                gl.bind_framebuffer(
+                    glow::READ_FRAMEBUFFER,
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
                    gl::READ_FRAMEBUFFER,
```

File: src/chains.rs [hunk 2]
```diff
@@ -191,23 +197,31 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
...
+                    frame_buffer(front_info.framebuffer_object),
+                );
+                debug_assert_eq!(gl.get_error(), glow::NO_ERROR);
+                gl.bind_framebuffer(
+                    glow::DRAW_FRAMEBUFFER,
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
                    gl::DRAW_FRAMEBUFFER,
```

File: src/chains.rs [hunk 3]
```diff
@@ -191,23 +197,31 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
...
-                gl::COLOR_BUFFER_BIT | gl::DEPTH_BUFFER_BIT | gl::STENCIL_BUFFER_BIT,
-                gl::NEAREST,
-            );
-            debug_assert_eq!(gl.get_error(), gl::NO_ERROR);
+            unsafe {
+                gl.bind_framebuffer(
+                    glow::READ_FRAMEBUFFER,
+                    frame_buffer(front_info.framebuffer_object),
+                );
+                debug_assert_eq!(gl.get_error(), glow::NO_ERROR);
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
                debug_assert_eq!(gl.get_error(), gl::NO_ERROR);
```

File: src/chains.rs [hunk 4]
```diff
@@ -191,23 +197,31 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
...
+                gl.bind_framebuffer(
+                    glow::DRAW_FRAMEBUFFER,
+                    frame_buffer(back_info.framebuffer_object),
+                );
+                debug_assert_eq!(gl.get_error(), glow::NO_ERROR);
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
                debug_assert_eq!(gl.get_error(), gl::NO_ERROR);
```

File: src/chains.rs [hunk 5]
```diff
@@ -338,7 +352,7 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
         &mut self,
         device: &mut Device,
         context: &mut Device::Context,
-        gl: &Gl,
+        gl: &glow::Context,
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
        gl: &gl::Context,
```

File: src/chains.rs [hunk 6]
```diff
@@ -564,7 +581,7 @@ impl<Device: DeviceAPI> SwapChain<Device> {
         &self,
         device: &mut Device,
         context: &mut Device::Context,
-        gl: &Gl,
+        gl: &glow::Context,
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
        gl: &gl::Context,
```

File: src/chains.rs [hunk 7]
```diff
@@ -51,7 +57,7 @@ struct SwapChainData<Device: DeviceAPI> {
 }
 
 pub enum PreserveBuffer<'a> {
-    Yes(&'a Gl),
+    Yes(&'a glow::Context),
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
    Yes(&'a gl::Context),
```

File: src/chains.rs [hunk 8]
```diff
@@ -26,14 +26,20 @@ use crate::device::Device as DeviceAPI;
...
 use std::collections::hash_map::Entry;
 use std::fmt::Debug;
 use std::hash::Hash;
 use std::mem;
+use std::num::NonZero;
 use std::sync::{Arc, Mutex, MutexGuard, RwLock, RwLockReadGuard, RwLockWriteGuard};
 
+fn frame_buffer(frame_buffer: u32) -> Option<glow::NativeFramebuffer> {
+    NonZero::new(frame_buffer).map(glow::NativeFramebuffer)
+}
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
fn frame_buffer(frame_buffer: u32) -> Option<gl::NativeFramebuffer> {
    NonZero::new(frame_buffer).map(gl::NativeFramebuffer)
}
```

File: src/chains.rs [hunk 9]
```diff
@@ -26,14 +26,20 @@ use crate::device::Device as DeviceAPI;
...
 use log::debug;
-use sparkle::gl::{self, GLuint, Gl};
 use std::collections::hash_map::Entry;
 use std::fmt::Debug;
 use std::hash::Hash;
 use std::mem;
+use std::num::NonZero;
 use std::sync::{Arc, Mutex, MutexGuard, RwLock, RwLockReadGuard, RwLockWriteGuard};
 
+fn frame_buffer(frame_buffer: u32) -> Option<glow::NativeFramebuffer> {
```
Comment (mrobinson) [inline src/chains.rs]: I think it makes sense to make this a `fn framebuffer()` getter on `impl Surface`. Note that `framebuffer` is spelled without the space in other parts of `surfman`.

**Review by mrobinson — CHANGES_REQUESTED** [review]

File: src/chains.rs (hunk 7 above)
Comment (sagudev) [reply src/chains.rs]: Actually `use glow::Context as Gl` will make even less changes.

Comment (mrobinson) [reply src/chains.rs]: I think it's okay to have a larger diff as long as things are consistent. 😁

File: src/chains.rs (hunk 9 above)
Comment (sagudev) [reply src/chains.rs]: > I think it makes sense to make this a fn framebuffer() getter on impl Surface

Er, what? glow is limited to chains.rs file and this function is only used to prepare input arguments for glow.

File: src/chains.rs [hunk 10]
```diff
@@ -26,14 +26,21 @@ use crate::device::Device as DeviceAPI;
...
 use std::collections::hash_map::Entry;
 use std::fmt::Debug;
 use std::hash::Hash;
 use std::mem;
+use std::num::NonZero;
 use std::sync::{Arc, Mutex, MutexGuard, RwLock, RwLockReadGuard, RwLockWriteGuard};
 
+fn framebuffer(framebuffer: u32) -> Option<gl::NativeFramebuffer> {
+    NonZero::new(framebuffer).map(gl::NativeFramebuffer)
+}
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
impl Surface {
    fn framebuffer(&self) -> Option<gl::NativeFramebuffer> {
        NonZero::new(self.framebuffer_object).map(gl::NativeFramebuffer)
    }
}
```

File: src/chains.rs [hunk 11]
```diff
@@ -191,23 +198,31 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
...
-                back_info.size.width,
-                back_info.size.height,
-                gl::COLOR_BUFFER_BIT | gl::DEPTH_BUFFER_BIT | gl::STENCIL_BUFFER_BIT,
-                gl::NEAREST,
-            );
-            debug_assert_eq!(gl.get_error(), gl::NO_ERROR);
+            unsafe {
+                gl.bind_framebuffer(
+                    gl::READ_FRAMEBUFFER,
+                    framebuffer(front_info.framebuffer_object),
```
Comment (mrobinson) [inline src/chains.rs]:
```suggestion
                    front_info.framebuffer(),
```

**Review by mrobinson — CHANGES_REQUESTED** [review]

File: src/chains.rs (hunk 10 above)
Comment (sagudev) [reply src/chains.rs]: We cannot remove all usages of framebuffer but https://github.com/servo/surfman/pull/318/commits/5d0d31ca6e20aad15b7ee6a277fe6ccfaddf4931 does most of it.

File: src/chains.rs [hunk 12]
```diff
@@ -26,14 +26,27 @@ use crate::device::Device as DeviceAPI;
...
 use std::collections::hash_map::Entry;
 use std::fmt::Debug;
 use std::hash::Hash;
 use std::mem;
+use std::num::NonZero;
 use std::sync::{Arc, Mutex, MutexGuard, RwLock, RwLockReadGuard, RwLockWriteGuard};
 
+fn framebuffer(framebuffer: u32) -> Option<gl::NativeFramebuffer> {
+    NonZero::new(framebuffer).map(gl::NativeFramebuffer)
+}
```
Comment (mrobinson) [inline src/chains.rs]: Let's move this into the function where it is used now that it is only used in one place.

Comment (sagudev) [reply src/chains.rs]: actually it's also used in `impl crate::SurfaceInfo { fn framebuffer(...` so we if we move it we will need to duplicate the code ...

File: src/chains.rs [hunk 13]
```diff
@@ -412,29 +428,31 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
...
-        gl.depth_mask(depth_mask[0] != 0);
-        gl.stencil_mask(stencil_mask[0] as gl::GLuint);
-        if scissor_enabled {
-            gl.enable(gl::SCISSOR_TEST);
-        }
-        if rasterizer_enabled {
-            gl.enable(gl::RASTERIZER_DISCARD);
+        unsafe {
+            gl.bind_framebuffer(gl::DRAW_FRAMEBUFFER, framebuffer(bound_fbos[0] as _));
+            gl.bind_framebuffer(gl::READ_FRAMEBUFFER, framebuffer(bound_fbos[1] as _));
```
Comment (sagudev) [inline src/chains.rs]: also used here

File: src/chains.rs (hunk 12 above)
Comment (mrobinson) [reply src/chains.rs]: I think it's okay to duplicate this code which is one line.

**Review by mrobinson — APPROVED** [review]

File: src/chains.rs [hunk 14]
```diff
@@ -351,17 +366,17 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
...
-            gl.get_integer_v(gl::STENCIL_WRITEMASK, &mut stencil_mask[..]);
-            gl.get_boolean_v(gl::COLOR_WRITEMASK, &mut color_mask[..]);
+            gl.get_parameter_i32_slice(gl::DRAW_FRAMEBUFFER_BINDING, &mut bound_fbos[0..]);
+            gl.get_parameter_i32_slice(gl::READ_FRAMEBUFFER_BINDING, &mut bound_fbos[1..]);
+            gl.get_parameter_f32_slice(gl::COLOR_CLEAR_VALUE, &mut clear_color[..]);
+            gl.get_parameter_f32_slice(gl::DEPTH_CLEAR_VALUE, &mut clear_depth[..]);
+            gl.get_parameter_i32_slice(gl::STENCIL_CLEAR_VALUE, &mut clear_stencil[..]);
+            gl.get_parameter_i32_slice(gl::DEPTH_WRITEMASK, &mut depth_mask[..]);
+            gl.get_parameter_i32_slice(gl::STENCIL_WRITEMASK, &mut stencil_mask[..]);
+            gl.get_parameter_i32_slice(gl::COLOR_WRITEMASK, &mut color_mask[..]);
```
Comment (sagudev) [inline src/chains.rs]: this is wrong, GLboolean != int so we need https://github.com/grovesNL/glow/pull/304

File: src/chains.rs (hunk 12 above)
Comment (sagudev) [reply src/chains.rs]: Actually we can use `get_parameter_framebuffer` (currently not yet release in glow) to eliminate last uses of framebuffer function.

Comment (mrobinson) [reply src/chains.rs]: That sounds perfect.

File: src/chains.rs [hunk 15]
```diff
@@ -344,24 +355,25 @@ impl<Device: DeviceAPI> SwapChainData<Device> {
...
-            gl.get_integer_v(gl::STENCIL_WRITEMASK, &mut stencil_mask[..]);
-            gl.get_boolean_v(gl::COLOR_WRITEMASK, &mut color_mask[..]);
+            draw_fbo = gl.get_parameter_framebuffer(gl::DRAW_FRAMEBUFFER_BINDING);
+            read_fbo = gl.get_parameter_framebuffer(gl::READ_FRAMEBUFFER_BINDING);
+            gl.get_parameter_f32_slice(gl::COLOR_CLEAR_VALUE, &mut clear_color[..]);
+            gl.get_parameter_f32_slice(gl::DEPTH_CLEAR_VALUE, &mut clear_depth[..]);
+            gl.get_parameter_i32_slice(gl::STENCIL_CLEAR_VALUE, &mut clear_stencil[..]);
+            depth_mask = gl.get_parameter_bool(gl::DEPTH_WRITEMASK);
+            gl.get_parameter_i32_slice(gl::STENCIL_WRITEMASK, &mut stencil_mask[..]);
+            gl.get_parameter_i32_slice(gl::COLOR_WRITEMASK, &mut color_mask[..]);
```
Comment (sagudev) [inline src/chains.rs]: We need https://github.com/grovesNL/glow/pull/308.

**Comment by sagudev** [comment]
This is breaking change and it will need version bump!

**Review by mrobinson — APPROVED** [review]

**Pull request merged**